### PR TITLE
✨RE:Add support for CAaja ads network 

### PR DESCRIPTION
**Pull request by matsuda-takehiro**
Adds support for CAaja ads network.

Hello.
We are A.J.A., a subsidiary of CyberAgent which provides ad-network service in Japan. (https://www.cyberagent.co.jp/) CA
(https://aja-kk.co.jp/) AJA
We'd like to add support to amp-ad.

Thanks.

previous↓
https://github.com/ampproject/amphtml/pull/20454

### CODE REVIEW

**Comment by mrjoro** [comment]
Sorry for the issues with the CLA, but I'm glad to see it's working now!

File: examples/ads.amp.html [hunk 1]
```diff
@@ -719,6 +720,13 @@ <h2>Bringhub Mini-Storefront</h2>
       heights="(max-width: 270px) 1280px, (max-width:553px) 640px, 338px">
   </amp-embed>
 
+  <h2>CA A.J.A. </h2>
+  <amp-ad width="300" height="250"
```
Comment (torch2424) [inline examples/ads.amp.html]: So it seems like you missed the attribute `widgetsIds` 😄 

screen shot 2019-02-04 at 12 10 05 pm

Other than that, everything looks great! 👍 Thanks for working with us and implementing all the changes!

**Comment by torch2424** [comment]
@matsuda-takehiro Thank you for making the changes!

So I went ahead and tested it again, but it seems like it doesn't load still:

screen shot 2019-02-05 at 2 13 20 pm

Is there something special I need to do?

Also, thank you very much for your patience, and working with out requests 😄

**Comment by matsuda-takehiro** [comment]
Thank you!　@torch2424 

Then, can you see the 16306 (AJA re-PR)?

I used <amp-iframe> by using AMP embed types. I'd like to use <amp-ad> from here.

I also tried to refer to 6478, but it did not work as well .....I got a percy error...


I apologize for the inconvenience😭



Should I listen with SLACK?..

**Comment by torch2424** [comment]
@matsuda-takehiro So I looked at the PR numbers you mentioned, and I am not quite sure what you mean by them. Also, not quite sure what you meant by AMP embed types. Could you elaborate a bit more? 😄 

But yes, the wg-ads channel is a great place to reach us as well 👍 

In terms of this PR, I think this is ready to be merged, I just want to be sure the ads will load before merging. So maybe it's a server side issue on your end?

Thanks! 😄

**Comment by matsuda-takehiro** [comment]
Thank you!　@torch2424 
I wanted to change from AMP-embed to AMP-ad.
I am sorry to complicate it.

Thank you! Maybe I will ask you a question with SLACK!

The number you added may be wrong. . .sorry...😭

Thank you for your time and consideration.


If you can give me advice, I would like to know how to change to AMP-ad. .(from16306)

**Comment by zhouyx** [comment]
@matsuda-takehiro So @torch2424 and I tried looking at #16306 and #6478

Looked like #16306 added `<amp-embed>` support for vendor type `aja`, #6478 added `<amp-ad>` support for vendor type `caajainfeed`. And this PR tries to add `<amp-ad>` support for `caaja`. 

Do you or your company own these three ad networks? If so could you please let us know how they relate to each other?  Thank you

**Comment by matsuda-takehiro** [comment]
Thank you!　@zhouyx 
Apparently, I heard that PR of another section of CA is 6478.

I do not know well with AJA. I'm sorry

16306 is AJA's one.
However, as the following problems occurred, I would like to be able to use the amp-ad format this time. . .

- The height is secured according to the fixed value, and the height is not expanded even if the frame of the widget increases.
- There is a rule that it is necessary to always set a fixed value.

Sorry for my lack of explanation😭

**Comment by zhouyx** [comment]
@matsuda-takehiro Thank you for the explanation. 

I am trying to understand the requirement here better. So `<amp-ad>` and `<amp-embed>` are two identical components. The only difference is that we apply default placeholder for `<amp-ad>` but not for `<amp-embed>`. Even without this PR, you can do `<amp-ad type="aja">` without any issue

For more information about amp-ad/amp-embed layout, please refer to the document [here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-ad/amp-ad.md). As you can find, "fill, fixed, fixed-height, flex-item, intrinsic, nodisplay, responsive" are all supported, and you can find more information about these layout types [here](https://www.ampproject.org/docs/design/responsive/control_layout). 

Given this, do you still see the need of adding the amp-ad support? Does the previous integration satisfy your requirement? Please let us know if there's anything I can help with. Thank you.

**Comment by matsuda-takehiro** [comment]
Thank you!　@zhouyx 

Thank you very much for giving me a hint.
I will discuss again.

Would you please wait for a while?

I'm sorry to trouble you.😭

**Comment by zhouyx** [comment]
@matsuda-takehiro I'm more than happy to help. Let me know once you have an update or if there's anything else I can help with. Thank you!

**Comment by matsuda-takehiro** [comment]
Thank you!　@zhouyx

Are there any attributes and values necessary to use amp-ad?
Is there anything other than <amp-ad type = "aja">?

I'm sorry if it was a rudimentary question😭

**Comment by matsuda-takehiro** [comment]
// ensure we have valid widgetIds value
   validateData (data, ['widgetids']);

Is this okay without it?↑

Please tell me what attributes and values you need in VendorName.js(aja.js)!

**Comment by matsuda-takehiro** [comment]
Thank you!　@zhouyx

I am sorry to complicate it. ↑ 2

I am in the direction to change aja.js.

Therefore, please let us know the attributes and values necessary to run amp-ad.

The current situation was used as follows.↓

```
<amp-iframe
            width="320"
            height="100"
            layout="responsive"
            sandbox="allow-scripts allow-same-origin allow-top-navigation allow-popups"
            frameborder="0"
            resizable
            src="https://static.aja-recommend.com/html/amp.html?ssp_code=XXXXXXXX&content_id=YYYY&kv.genre=other">
            <div overflow tabindex=0 role=button aria-label="Read more" class="ampReadmore">Read more</div>
            <amp-img layout="fill" src="https://static.aja-recommend.com/amp/transpanret.gif" placeholder></amp-img>
        </amp-iframe>
```

for that reason,
* The height is secured according to the fixed value, and the height is not expanded even if the frame of the widget increases.
* There is a rule that it is necessary to always set a fixed value.

The problem like the one above has come out.
So, I'd like to use amp - ad,

**Comment by matsuda-takehiro** [comment]
image

It looks like this screenshot.

In aja.js
// ensure we have valid widgetIds value
  validateData (data, ['widgetids']);
There are such sentences, but widgetids do not seem to be using AJA.

Please tell me what attributes and values you need in VendorName.js(aja.js).

**Comment by matsuda-takehiro** [comment]
https://www.ampproject.org/docs/reference/components/amp-ad
I looked over here! ↑

**Comment by zhouyx** [comment]
@matsuda-takehiro `widgetIds` is a required parameter by the already merge [aja network](https://github.com/ampproject/amphtml/blob/master/ads/aja.md#parameters).

If I understand it correctly, the widgetId is passed to the `global._aja` object and will by used by your aja AMP JS script. Is that correct?

**Comment by matsuda-takehiro** [comment]
Thank you!　@zhouyx

I think it is probably a mistake.
What should I do?

Should I create a new amp-ad again?

```
width="320"
height="100"
layout="responsive"
sandbox="allow-scripts allow-same-origin allow-top-navigation allow-popups"
frameborder="0"
resizable
src="https://static.aja-recommend.com/html/amp.html?ssp_code=XXXXXXXX&content_id=YYYY&kv.genre=other">
<div overflow tabindex=0 role=button aria-label="Read more" class="ampReadmore">Read more</div>
<amp-img>
```
I only use this attribute.

**Comment by matsuda-takehiro** [comment]
Tanaka says that he mimicked PR of outbrain.

In my case, since "js" read with "load script" is different from "caajainfeed", this time I imitated aja

**Comment by zhouyx** [comment]
@matsuda-takehiro I found that #16306 is authored by @hideaki0817 

Can you confirm with them that widgetId is not required and remove it from the current `aja.js` implementation in AMP runtime? Thank you.

**Comment by matsuda-takehiro** [comment]
Thank you very much for your early reply! @zhouyx

Oops! Tanaka is about hideaki0817
sorry.


Yes. Please.
However, I do not know if I will be able to display it if I erase it, I will cause you trouble. I'm sorry.

**Comment by matsuda-takehiro** [comment]
Could you please remove it? Or should I do?

We are sorry for the inconvenience.😭

**Comment by zhouyx** [comment]
@matsuda-takehiro I see, you are welcome to create PR to remove the `widgetId` requirement then, as long as it gets code owner approval. In this case @hideaki0817 

I want to make sure that this won't affect pages that has `<amp-embed type="aja">`. And based on the example you put above. 
`iframe.src = '......ssp_code=XXXXXXXX&content_id=YYYY``, do you expect the `ssp_code` and `content_id` to be required for the ad? Thank you

**Comment by matsuda-takehiro** [comment]
Thank you!　@zhouyx

I received permission from @hideaki0817
I'd like to revise this PR because I do not know if it has any impact, but what about it?

Although I use ssp_code, I feel that I do not use the content_id in particular.

**Comment by zhouyx** [comment]
@matsuda-takehiro based on our discussion, I also think it makes sense to revise this PR. I believe we agreed on several things.
1. Reuse the current aja integration, not introducing a new ad vendor type.
2. Remove the requirement for `widgetIds` from aja implmentation (Will need @hideaki0817 approval for the PR)
3. We agree that the change to aja integration won't break existing usage. 
4. Add support to `ssp_code`. (Is it a required param or not?)

Please let me know if this makes sense, and if I'm missing anything. If everything looks correct, feel free to make the change. Thank you.

**Comment by matsuda-takehiro** [comment]
Thank you very much! !@zhouyx！

As a result of consulting within the company, it turned out that even if you changed the file produced by@hideaki0817　, it does not affect AMP currently in use!🏃😆

I hope you can help us again after the change!🙇

**Comment by matsuda-takehiro** [comment]
Are there any elements necessary for "/ads/yournetwork.js"?

**Comment by matsuda-takehiro** [comment]
image

git fetch upstream master→error 
"Please make sure you have the correct access rights
and the repository exists."

I was told that there is no authority. How can I do it?

**Comment by zhouyx** [comment]
@matsuda-takehiro 
>Are there any elements necessary for "/ads/yournetwork.js"?
Please let me know if you have a model!

Please refer to our doc [here](https://github.com/ampproject/amphtml/blob/master/ads/README.md), specifically this [section](https://github.com/ampproject/amphtml/blob/master/ads/README.md#developer-guidelines-for-a-pull-request) on how to create a PR. You can also search for our example ad type `_ping_` in the AMP codebase to get our fake ad network example. 

Regarding the git question. Can I know what does `git remote -v` returns? Thanks

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx
image

It looks like this image...😭

**Comment by zhouyx** [comment]
Thank you @matsuda-takehiro 
That looks correct to me. Did you connect to GitHub with https or ssh? If you clone with ssh, could you also make sure that the ssh key is generated correctly. 
https://help.github.com/en/articles/set-up-git

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx
It is probably https...

**Comment by zhouyx** [comment]
Thanks for the information. I think the issue is that the remote urls are ssh urls. If you're connecting with https, please switch to https url instead. 
https://help.github.com/en/articles/changing-a-remotes-url#switching-remote-urls-from-ssh-to-https

**Comment by matsuda-takehiro** [comment]
image

!!!!

Thank you very much! !@zhouyx！

It looks like it is going ahead！

sorry for making you go to all this trouble...

**Comment by matsuda-takehiro** [comment]
image

 today 10:01

Thank you very much! !

**Comment by matsuda-takehiro** [comment]
/ads/yournetwork.js　(aja.js or caaja.js）

Can caaja.js(↑) 　div and script be inserted by caaja.js? If caaja.js can,

- Can caaja.js insert it as a sibling node of amp-ad?
- Can caaja.js insert in an iframe in amp-ad?

**Comment by matsuda-takehiro** [comment]
How can I do if I want to test before issuing PR?
Or maybe not?
I did not understand it even if I read here（https://github.com/ampproject/amphtml/blob/master/ads/README.md#files-to-change）

**Comment by zhouyx** [comment]
@matsuda-takehiro Glad it works. 

>Can caaja.js insert in an iframe in amp-ad?
AMP runtime creates an iframe for every `<amp-ad>` component, and loads the ad network script within the iframe. 
And with your ad network script, you can do things from creating div to load some other script within the iframe. 

If you want to test the change locally, you can run our local server https://github.com/ampproject/amphtml/blob/master/contributing/getting-started-quick.md#build-amp--run-a-local-server
And then go to `localhost:8000/examples/ads.amp.html?type={adTypeName}` to see your examples.

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx　！

Does the test also reflect "/ads/yournetwork.js　(aja.js or caaja.js）" changes?
Programming of "/ads/yournetwork.js　(aja.js or caaja.js）" seems to be gonna fumble ...😭
However, if I can do a test including "/ads/yournetwork.js" it seems it will work! 😆

**Comment by zhouyx** [comment]
@matsuda-takehiro When you run the local server, the locally built ad script will be used. As said that should reflect any change to `ads/yournetwork.js`. You can always print something out to verify : ) Let me know how it goes.

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

I can not use yarn(?) yet.
So I will be using yarn.

**Comment by matsuda-takehiro** [comment]
Hmm
Will it fail if I use MAMP?

**Comment by matsuda-takehiro** [comment]
別件ですが、以下のようにsrcをamp-adのiframe内で読み込む方法はございますか？

```
  <amp-iframe
            width="320"
            height="100"
            layout="responsive"
            sandbox="allow-scripts allow-same-origin allow-top-navigation allow-popups"
            frameborder="0"
            resizable
            src="https://static.aja-recommend.com/html/amp.html?ssp_code=XXXXXX&content_id=YYYY&kv.genre=other">
            <div overflow tabindex=0 role=button aria-label="Read more" class="ampReadmore">Read more</div>
            <amp-img layout="fill" src="https://static.aja-recommend.com/amp/transpanret.gif" placeholder></amp-img>
        </amp-iframe>
```
```
 src="https://static.aja-recommend.com/html/amp.html?ssp_code=XXXXXX&content_id=YYYY&kv.genre=other">
```

Like writeScript and loadScript

loadScript(global, 'https://static.aja-recommend.com/html/amp.html?ssp_code=XXXXX');

**Comment by zhouyx** [comment]
@matsuda-takehiro Let me guess, you want to dynamically calculate the src. What I would do this
```
<amp-ad type='aja'
   width=...
   height=...
   data-ssp-code=xxx>
</amp-ad>
```

Then in your `aja.js` script, calculate the url based on `data-ssp-code` value
```
url = `https://static.aja-recommend.com/html/amp.html` + encodeURIComponent(data['sspCode'])`
```

and then load the script
```
loadScript(url)
```

Let me know if that works for you.

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx

Should I use Camel Case?　（sspCode）
OK. Thank you



Although I have not finished it yet at all, I think that I will make use of the advice I received promptly!

image

**Comment by matsuda-takehiro** [comment]
I came up with 2 patterns!
- Load amp.html
   - The way you taught me
- Load CDN
   - TODO　
```
  //CDN case
    //var el = document.createElement('script');
    //el.setAttribute('src', 'https://cdn.as.amanad.adtdp.com/tag/' + ZZZZZ + '.js');
    //el.setAttribute('data-aja-ssp-asi', SSPCODE);
    //el.setAttribute('async', 'true');
  //Completion example→  <script src="https://cdn.as.amanad.adtdp.com/tag/ZZZZZ.js" data-aja-ssp-asi="SSPCODE" async="true"></script>
```

**Comment by matsuda-takehiro** [comment]
Can I build another pull request and fix aja.js or caaja.js?

Sorry for turning over two times

**Comment by zhouyx** [comment]
>Should I use Camel Case?　（sspCode）

AMP handles data attribute in a way to convert `data-ssp-code` to `data['sspCode']`, and `data-sspCode` to `data['sspcode']`. So I wouldn't recommend using camel case because `data-sspCode` is essentially the same as `data-sspcode`. 

And please refer to our helper method `writeScript` and `loadScript` https://github.com/ampproject/amphtml/blob/master/3p/3p.js#L84, maybe they will be helpful when trying to load script. 

>Can I build another pull request and fix aja.js or caaja.js?

I'm fine with that. Please refer to this PR in your new PR.

**Comment by matsuda-takehiro** [comment]
Thank you @zhouyx 
I will read https://github.com/ampproject/amphtml/blob/master/3p/3p.js#L84 at once!

**Comment by matsuda-takehiro** [comment]
#21067

**Pull request closed without merging**